vividus-plugin-mobile-app/src/main/java/org/vividus/bdd/mobileapp/steps/ActionSteps.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2019-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.bdd.mobileapp.steps;
+
+import java.time.Duration;
+
+import org.jbehave.core.annotations.When;
+import org.vividus.bdd.steps.ui.validation.IBaseValidations;
+import org.vividus.mobileapp.action.TapActions;
+import org.vividus.ui.action.search.Locator;
+
+public class ActionSteps
+{
+    private final TapActions tapActions;
+    private final IBaseValidations baseValidations;
+
+    public ActionSteps(TapActions tapActions, IBaseValidations baseValidations)
+    {
+        this.tapActions = tapActions;
+        this.baseValidations = baseValidations;
+    }
+
+    /**
+     * Taps on <b>element</b> located by <b>locator</b> with specified <b>duration</b>
+     * @param locator locator to find an element
+     * @param duration between an element is pressed and released in
+     * <a href="https://en.wikipedia.org/wiki/ISO_8601">ISO_8601</a> format
+     */
+    @When("I tap on element located by `$locator` with duration `$duration`")
+    public void tapByLocator(Locator locator, Duration duration)
+    {
+        baseValidations.assertElementExists("The element to tap", locator).ifPresent(e -> tapActions.tap(e, duration));
```

Comment:
Since the tap will be called only if the element exists why the tap performs a nullability check?

Reply:
it's not really nullability check, it's just optional api which let us to inline tap invocation

Reply:
You're checking element in TapActions

Reply:
lin 43

Reply:
In my understanding we have two layers of API, the first one - are steps for users, and the second one - for internal use, that's why I put null check into TapActions because it can be used in other places

Reply:
Yep, and since API you're using provides null-safety no reason to use an optional?

Reply:
base validation - is not the only way to get elements